### PR TITLE
Recurrent batch normalization

### PR DESCRIPTION
**Pull request by elvinpoon**
http://arxiv.org/abs/1603.09025
Flag 'batch_norm' added to BasicLSTMCell to control whether to apply BN to hidden states

### CODE REVIEW

File: tflearn/layers/recurrent.py [hunk 1]
```diff
@@ -22,6 +22,7 @@
 from .. import activations
 from .. import initializations
 from .. import variables as va
+from normalization import batch_normalization
```
Comment (aymericdamien) [inline tflearn/layers/recurrent.py]: it should be:
`from .normalization import batch_normalization`

File: tflearn/layers/recurrent.py [hunk 2]
```diff
@@ -514,7 +515,13 @@ def __call__(self, inputs, state, scope=None):
             new_c = (c * self._inner_activation(f + self._forget_bias) +
                      self._inner_activation(i) *
                      self._activation(j))
-            new_h = self._activation(new_c) * self._inner_activation(o)
+            
+            # hidden-to-hidden batch normalizaiton
+            if batch_norm == False:
```
Comment (aymericdamien) [inline tflearn/layers/recurrent.py]: here you have an error, you have to save batch_norm under self.batch_norm

File: tflearn/layers/recurrent.py [hunk 3]
```diff
@@ -514,7 +515,13 @@ def __call__(self, inputs, state, scope=None):
...
+                new_h = self._activation(new_c) * self._inner_activation(o)
+            else:
+                batch_norm_new_c = batch_normalization(new_c, restore=False)
```
Comment (aymericdamien) [inline tflearn/layers/recurrent.py]: here you can propagate arguments to batch_normalization: `trainable=self.trainable`, `restore=self.restore`, `reuse=self.reuse`, so that it can works synchronously with the Weights same parameters.

**Comment by aymericdamien** [comment]
That looks cool! I added some comments, so you can fix your code and I will merge it. Thanks :)

File: tflearn/layers/recurrent.py [hunk 4]
```diff
@@ -514,7 +515,13 @@ def __call__(self, inputs, state, scope=None):
...
+                new_h = self._activation(batch_norm_new_c) * self._inner_activation(o)
```
Comment (tgallice) [inline tflearn/layers/recurrent.py]: Just a comment about style, this can be rewrite like that to be more readable. 

```Python
if self.batch_norm == True:
    new_c = batch_normalization(new_c, trainable=self.trainable, restore=self.restore, reuse=self.reuse)

new_h = self._activation(new_c) * self._inner_activation(o)
```

**Comment by elvinpoon** [comment]
Thanks for your comment!

**Comment by tgallice** [comment]
In the paper they give some default values for **beta: 0** and **gamma: 0.1** (by default it's 1.0) This is maybe a good idea to use it ?

They also normalize the recurrent and input terms, add a normalization op in the `_linear` method can do the tricks.
What do you think ?

File: tflearn/layers/recurrent.py (hunk 4 above)
Comment (cooijmanstim) [inline tflearn/layers/recurrent.py]: This changes the meaning though, since `new_c` is a return value. Previously the returned cell value was not normalized (as cf the paper it shouldn't be), now it is.

**Comment by elvinpoon** [comment]
Initializing gamma to 0.1 is probably a good idea, I will use it in later commit. I think we should instead set bias to constant zero because you only need to modify recurrent.py. It's simpler.
As for recurrent and input terms, I have the plan to implement normalization in _linear. But there is one paper that claims adding BN for input term will hurt generalization performance. 
See http://arxiv.org/abs/1510.01378
It says "We also show that when applied to the input-to-hidden transitions, batch normalization can lead to a faster convergence of the training criterion but doesn't seem to improve the generalization performance on both our language modelling and speech recognition tasks. "

 Adding BN to both will increase computation. Probably we should just try to apply BN to recurrent terms. But I'm not sure about this.
What do you think?

**Comment by cooijmanstim** [comment]
At the time of http://arxiv.org/abs/1510.01378 it wasn't understood that initializing gamma to 1 kills the gradient through the tanh, so that result is suspect. The first author on that paper is also on the newer paper (http://arxiv.org/abs/1603.09025).

**Comment by elvinpoon** [comment]
Cool. Let's do this.

**Comment by elvinpoon** [comment]
I've applied BN to recurrent and input terms, and i've adjusted the gamma weight. Maybe you wanna take a look and see if everything is ok.

**Comment by aymericdamien** [comment]
Great job! Let me review it :)

**Comment by aymericdamien** [comment]
It is working great! Thanks @elvinpoon 
Later, we can also think about adding batch_norm to SimpleRNNCell and GRUCell, as well as lstm, gru and simple_rnn layers.
